m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,831 @@
+            bucketValFmt = String.format("%%.%df", builder.histogramBucketTagPrecision);
+
+            metricQueue = new LinkedBlockingQueue<>(builder.maxQueueSize);
+
+            addAndRunProcessor();
+        } catch (TTransportException | SocketException e) {
+            throw new IllegalArgumentException("Exception creating M3Reporter", e);
+        }
+    }
+
+    private int calculateFreeBytes(int maxPacketSizeBytes, Set<MetricTag> commonTags) {
+        metricBatch.setCommonTags(commonTags);
+        metricBatch.setMetrics(new ArrayList<Metric>());
+
+        int size;
+
+        try {
+            metricBatch.write(calcProtocol);
+            size = calc.getSizeAndReset();
+        } catch (TException e) {
+            LOG.warn("Unable to calculate metric batch size. Defaulting to: %s", DEFAULT_METRIC_SIZE);
+            size = DEFAULT_METRIC_SIZE;
+        }
+
+        int numOverheadBytes = EMIT_METRIC_BATCH_OVERHEAD + size;
+
+        freeBytes = maxPacketSizeBytes - numOverheadBytes;
+
+        if (freeBytes <= 0) {
+            throw new IllegalArgumentException("Common tags serialized size exceeds packet size");
+        }
+
+        return freeBytes;
+    }
+
+    private static String getHostName() {
+        try {
+            return InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            LOG.warn("Unable to determine hostname. Defaulting to: %s", DEFAULT_TAG_VALUE);
+            return DEFAULT_TAG_VALUE;
+        }
+    }
+
+    private void addAndRunProcessor() {
+        phaser.register();
+
+        executor.execute(new Processor());
+    }
+
+    @Override
+    public Capabilities capabilities() {
+        return CapableOf.REPORTING_TAGGING;
+    }
+
+    @Override
+    public void flush() {
+        try {
+            metricQueue.put(SizedMetric.FLUSH);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void flush(List<Metric> metrics) {
+        flush(metrics, DEFAULT_FLUSH_TRY_COUNT);
+    }
+
+    private void flush(List<Metric> metrics, int maxTries) {
+        if (metrics.isEmpty()) {
+            return;
+        }
+
+        TException exception = null;
+
+        int tries = 0;
+        boolean tryFlush = true;
+
+        synchronized (metricBatch) {
+            metricBatch.setMetrics(metrics);
+
+            while (tryFlush) {
+                try {
+                    client.emitMetricBatch(metricBatch);
+
+                    tryFlush = false;
+                } catch (TException tException) {
+                    tries++;
+
+                    if (tries < maxTries) {
+                        try {
+                            Thread.sleep(FLUSH_RETRY_DELAY_MILLIS);
+                        } catch (InterruptedException interruptedException) {
+                            Thread.currentThread().interrupt();
```

Comment:
why do we need to do this?

Reply:
This is a good practice when catching `InterruptedException`s, especially when not really handling them in a meaningful way. Catching such exceptions turns off the "interrupted flag", so logic further in the code path won't know that someone tried to interrupt the thread. Here, we set that flag back just in case the thread elsewhere wants to handle it some other way later.

Reply:
Is this thrown per-thread? If so, shouldn't the bg-processing thread handle it once, and never have to notify anyone after?

Reply:
Agreed. Since we want to simply log and continue processing most of the time, we should not re-interrupt. Removing these.